src/interface/sparse_pack.hpp

```diff
@@ -0,0 +1,282 @@
+      : parthenon::variables::base_t<true>(std::forward<Ts>(args)...) {}
+  static std::string name() { return ".*"; }
+};
+} // namespace variables
+
+template <class... Ts>
+class SparsePack : public SparsePackBase {
+ public:
+  SparsePack() = default;
+
+  explicit SparsePack(const SparsePackBase &spb) : SparsePackBase(spb) {}
+
+  template <class T>
+  static SparsePack Make(T *pmd, const std::vector<MetadataFlag> &flags = {},
+                         bool fluxes = false, bool coarse = false) {
+    auto &cache = pmd->GetSparsePackCache();
+    return SparsePack(cache.Get(
+        pmd, PackDescriptor(std::vector<std::string>{Ts::name()...},
+                            std::vector<bool>{Ts::regex()...}, flags, fluxes, coarse)));
+  }
+
+  template <class T, class VAR_VEC>
+  static std::tuple<SparsePack, SparsePackIdxMap>
+  Make(T *pmd, const VAR_VEC &vars, const std::vector<MetadataFlag> &flags = {},
+       bool fluxes = false, bool coarse = false) {
+    static_assert(sizeof...(Ts) == 0);
+    auto out = SparsePackBase::Make(pmd, vars, flags, fluxes, coarse);
+    return {SparsePack(std::get<0>(out)), std::get<1>(out)};
+  }
+
+  template <class T>
+  static SparsePack MakeWithFluxes(T *pmd, const std::vector<MetadataFlag> &flags = {}) {
+    const bool coarse = false;
+    const bool fluxes = true;
+    return Make(pmd, flags, fluxes, coarse);
+  }
+
+  template <class T, class VAR_VEC>
+  static std::tuple<SparsePack, SparsePackIdxMap>
+  MakeWithFluxes(T *pmd, const VAR_VEC &vars,
+                 const std::vector<MetadataFlag> &flags = {}) {
+    const bool coarse = false;
+    const bool fluxes = true;
+    Make(pmd, vars, flags, fluxes, coarse);
+  }
+
+  template <class T>
+  static SparsePack MakeWithCoarse(T *pmd, const std::vector<MetadataFlag> &flags = {}) {
+    const bool coarse = true;
+    const bool fluxes = false;
+    return Make(pmd, flags, fluxes, coarse);
+  }
+
+  template <class T, class VAR_VEC>
+  static std::tuple<SparsePack, SparsePackIdxMap>
+  MakeWithCoarse(T *pmd, const VAR_VEC &vars,
+                 const std::vector<MetadataFlag> &flags = {}) {
+    const bool coarse = true;
+    const bool fluxes = false;
+    Make(pmd, vars, flags, fluxes, coarse);
+  }
+
+  KOKKOS_FORCEINLINE_FUNCTION
+  int GetNBlocks() const { return nblocks_; }
+
+  KOKKOS_FORCEINLINE_FUNCTION
+  int GetNDim() const { return ndim_; }
+
+  KOKKOS_FORCEINLINE_FUNCTION
+  int GetDim(const int i) const {
+    assert(i > 0 && i < 6);
+    PARTHENON_REQUIRE(
+        i != 2, "Should not ask for the second dimension since it is logically ragged");
+    return dims_[i];
+  }
+
+  KOKKOS_INLINE_FUNCTION
+  const Coordinates_t &GetCoordinates(const int b) const { return coords_(b)(); }
+
+  // Bound overloads
+  KOKKOS_INLINE_FUNCTION int GetLowerBound(const int b) const { return bounds_(0, b, 0); }
+
+  KOKKOS_INLINE_FUNCTION int GetUpperBound(const int b) const {
+    return bounds_(1, b, nvar_ - 1);
+  }
```

Comment:
Should these have a default `b = 0`? That way when on a single meshblock, the user can neglect the block index

Reply:
Yeah, I think that makes sense although maybe we want to add an assert in that case the `nblock==0`

Reply:
On second though, I think allowing for default `b=0` basically double the number of operator overloads in `SparsePack` and make things much less readable. For now, lets force the user to explicitly write the block index. If it becomes clear that we need these overloads, we can add them in a future PR.

Reply:
Is this actually true? Why not just add a default argument, rather than adding an overload?

Reply:
I think for these three you can add just the default argument, but I think the interface should be consistent so that if you can elide the block number here you should be able to do it in all of the `operator()` methods.

Reply:
Ah, I see. Yes that adds some additional complexity. I think that would be nice. But maybe not necessary for now.